# Proceso de Extracción, Transformación y Carga de datos (ETL) 👨🏽‍💻 👩🏽‍💻

En este notebook, se cargará los datasets y se revisarán los tipos de cada columna para determinar el correcto. En el notebook también se verificará la presencia de registros duplicados y si algún registro en una columna tiene valores vacíos o nulos. 

Además de estas transformaciones, se importarán las bibliotecas que se utilizarán a lo largo del notebook para manipular los datos, incluido nuestro módulo personalizado llamado 'Funciones'. Finalmente, los datos procesados se exportarán y estarán listos para el análisis.

## Importamos las librerías necesarias 📚

In [5]:
import pandas as pd
import Funciones as F
import warnings
warnings.filterwarnings('ignore')
import sys
sys.path.insert(0, '../')

## Lectura de archivos 📂🔄

In [7]:
customers = pd.read_excel('../Datasets/bc_customers_dataset.xlsx')
customers

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  
0                         14409                 franca             SP  
1                          9790  sao bernardo do campo             SP  
2                          1151              sao paulo             SP  
3                          8775        mogi das cruzes             SP  
4                         13056               campinas             SP  
...                         ...                    ...            ...  
99436                      3937              sao paulo             SP  
99437                      6764        taboao da serra             SP  
99438                     60115              fortaleza             CE  
99439                     92120                 canoas             RS  
99440                      6703                  cotia             SP  

[99441 rows x 5 columns]

In [8]:
orders = pd.read_excel('../Datasets/bc_orders_dataset.xlsx')
orders

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99276  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99277  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99278  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99279  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99280  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status   order_approved_at order_received_by_local_operator_date  \
0        delivered 2017-10-02 11:07:15                   2017-10-04 19:55:00   
1        delivered 2018-07-26 03:24:27                   2018-07-26 14:31:00   
2        delivered 2018-08-08 08:55:23                   2018-08-08 13:50:00   
3        delivered 2017-11-18 19:45:59                   2017-11-22 13:39:59   
4        delivered 2018-02-13 22:20:29                   2018-02-14 19:46:34   
...            ...                 ...                                   ...   
99276    delivered 2017-03-09 09:54:05                   2017-03-10 11:18:03   
99277    delivered 2018-02-06 13:10:37                   2018-02-07 23:22:42   
99278    delivered 2017-08-27 15:04:16                   2017-08-28 20:52:26   
99279    delivered 2018-01-08 21:36:21                   2018-01-12 15:35:03   
99280    delivered 2018-03-09 11:20:28                   2018-03-09 22:11:59   

      order_delivered_to_customer_date order_delivery_promise_date  
0                  2017-10-10 21:25:13                  2017-10-18  
1                  2018-08-07 15:27:45                  2018-08-13  
2                  2018-08-17 18:06:29                  2018-09-04  
3                  2017-12-02 00:28:42                  2017-12-15  
4                  2018-02-16 18:17:02                  2018-02-26  
...                                ...                         ...  
99276              2017-03-17 15:08:01                  2017-03-28  
99277              2018-02-28 17:37:56                  2018-03-02  
99278              2017-09-21 11:24:17                  2017-09-27  
99279              2018-01-25 23:32:54                  2018-02-15  
99280              2018-03-16 13:08:30                  2018-04-03  

[99281 rows x 7 columns]

----

## Transformación de datos 🕵🏼‍♂️

### **Dataset Clientes**

Utilizaremos una función de nuestro modelo de 'Funciones' para revisar los tipos de datos de cada columna y detectar los valores nulos en ambos datasets.

In [9]:
F.analizar_datos(customers)

Nombre Tipos de Datos Únicos  % de Valores No Nulos  \
0               customer_id       [<class 'str'>]                  100.0   
1        customer_unique_id       [<class 'str'>]                  100.0   
2  customer_zip_code_prefix       [<class 'int'>]                  100.0   
3             customer_city       [<class 'str'>]                  100.0   
4            customer_state       [<class 'str'>]                  100.0   

   % de Valores Nulos  Cantidad de Valores Nulos  
0                 0.0                          0  
1                 0.0                          0  
2                 0.0                          0  
3                 0.0                          0  
4                 0.0                          0

In [10]:
F.analizar_datos(orders)

Nombre  \
0                               order_id   
1                            customer_id   
2                           order_status   
3                      order_approved_at   
4  order_received_by_local_operator_date   
5       order_delivered_to_customer_date   
6            order_delivery_promise_date   

                               Tipos de Datos Únicos  % de Valores No Nulos  \
0                                    [<class 'str'>]                 100.00   
1                                    [<class 'str'>]                 100.00   
2                                    [<class 'str'>]                 100.00   
3  [<class 'pandas._libs.tslibs.timestamps.Timest...                 100.00   
4  [<class 'pandas._libs.tslibs.timestamps.Timest...                  98.35   
5  [<class 'pandas._libs.tslibs.timestamps.Timest...                  97.16   
6  [<class 'pandas._libs.tslibs.timestamps.Timest...                 100.00   

   % de Valores Nulos  Cantidad de Valores Nulos  
0                0.00                          0  
1                0.00                          0  
2                0.00                          0  
3                0.00                          0  
4                1.65                       1637  
5                2.84                       2819  
6                0.00                          0

### ``Customer_id``

No se observan duplicados ni valores nulos en esta columna.

In [11]:
customers['customer_id'].duplicated().value_counts()

customer_id
False    99441
Name: count, dtype: int64

### ``Customer_unique_id``

Se observan 3345 valores duplicados en esta columna. Revisaremos más a fondo para verificar estos datos.

In [12]:
customers['customer_unique_id'].duplicated().value_counts()

customer_unique_id
False    96096
True      3345
Name: count, dtype: int64

Calculamos cuánto representan estos duplicados del total para evaluar si se pueden eliminar o no del conjunto de datos.

In [13]:
Porcentaje_unique = (3345*100/len(customers))
print(F'El porcentaje que representa de duplicados en la columna customer_unique_id es: {round(Porcentaje_unique,2)} del total de registros')

El porcentaje que representa de duplicados en la columna customer_unique_id es: 3.36 del total de registros


Se verifica que los registros duplicados realmente tengan los demás valores duplicados, lo cual se confirma revisando el DataFrame a continuación.

In [14]:
duplicados_unique = customers[customers.duplicated(subset=['customer_unique_id'], keep=False)]
duplicados_unique_ordenados = duplicados_unique.sort_values(by='customer_unique_id')
duplicados_unique_ordenados

customer_id                customer_unique_id  \
35608  24b0e2bd287e47d54d193e7bbb51103f  00172711b30d52eea8b313a7f2cced02   
19299  1afe8a9c67eec3516c09a8bdcc539090  00172711b30d52eea8b313a7f2cced02   
20023  1b4a75b3478138e99902678254b260f4  004288347e5e88a27ded2bb23747066c   
22066  f6efe5d5c7b85e12355f9d5c3db46da2  004288347e5e88a27ded2bb23747066c   
72451  49cf243e0d353cd418ca77868e24a670  004b45ec5c64187465168251cd1c9c2f   
...                                 ...                               ...   
75057  1ae563fdfa500d150be6578066d83998  ff922bdd6bafcdf99cb90d7f39cea5b3   
27992  bec0bf00ac5bee64ce8ef5283051a70c  ff922bdd6bafcdf99cb90d7f39cea5b3   
79859  d064be88116eb8b958727aec4cf56a59  ff922bdd6bafcdf99cb90d7f39cea5b3   
64323  4b231c90751c27521f7ee27ed2dc3b8f  ffe254cc039740e17dd15a5305035928   
12133  0088395699ea0fcd459bfbef084997db  ffe254cc039740e17dd15a5305035928   

       customer_zip_code_prefix customer_city customer_state  
35608                     45200        jequie             BA  
19299                     45200        jequie             BA  
20023                     26220   nova iguacu             RJ  
22066                     26220   nova iguacu             RJ  
72451                     57055        maceio             AL  
...                         ...           ...            ...  
75057                     17340  barra bonita             SP  
27992                     17340  barra bonita             SP  
79859                     17340  barra bonita             SP  
64323                     37640       extrema             MG  
12133                     37640       extrema             MG  

[6342 rows x 5 columns]

Después de un análisis exhaustivo de los datos, se ha determinado que un 3,36% de los registros presentan duplicados en esta columna.

Al ser un porcentaje bajo del total, he decidido eliminarlos, ya que la presencia de duplicados puede generar inconsistencias en el análisis y afectar la precisión de los resultados. Eliminar estos duplicados contribuirá a mantener la coherencia y la fiabilidad de los datos.

In [15]:
customers = customers.drop_duplicates(subset='customer_unique_id')

Revisamos la longitud del conjunto de datos nuevamente para corroborar que se eliminaron los duplicados. Efectivamente, se realizó la modificación.

In [16]:
len(customers)

96096

### ``customer_zip_code_prefix``

Se verificó que hay duplicados en esta columna, lo que nos indica que tenemos varios clientes de la misma zona.

In [17]:
customers['customer_zip_code_prefix'].duplicated().value_counts()

customer_zip_code_prefix
True     81114
False    14982
Name: count, dtype: int64

### ``customer_city``

Elegiremos los primeros 10 valores, ya que contamos con 4118 ciudades que se repiten, lo cual no tiene una representación significativa en porcentaje como las primeras 10 ciudades.

In [18]:
top_10_ciudades = customers['customer_city'].value_counts().head(10)
top_10_ciudades

customer_city
sao paulo                14971
rio de janeiro            6611
belo horizonte            2671
brasilia                  2066
curitiba                  1462
campinas                  1397
porto alegre              1326
salvador                  1208
guarulhos                 1151
sao bernardo do campo      908
Name: count, dtype: int64

Se corrigió el nombre de la columna a 'Customer_municipality', dado que se averiguó que en Brasil no se les denomina ciudades, sino municipios.

In [19]:
customers.rename(columns={'customer_city': 'customer_municipality'}, inplace=True)

### ``customer_state``

Se verifican los duplicados en los estados de Brasil, dado que el contador solo puede dar como máximo 27, que es la cantidad exacta de estados en Brasil.

In [20]:
customers['customer_state'].duplicated().value_counts()

customer_state
True     96069
False       27
Name: count, dtype: int64

Contamos cuántos clientes tenemos por estado para determinar cuál es el más relevante.

In [21]:
customers['customer_state'].value_counts()

customer_state
SP    40295
RJ    12377
MG    11255
RS     5277
PR     4882
SC     3529
BA     3276
DF     2073
ES     1963
GO     1951
PE     1604
CE     1311
PA      949
MT      875
MA      725
MS      693
PB      519
PI      482
RN      474
AL      401
SE      341
TO      273
RO      239
AM      143
AC       77
AP       67
RR       45
Name: count, dtype: int64

A modo de verificar la consistencia de los datos, pregunté a ChatGPT sobre la cantidad de ciudades y estados que posee Brasil. En el proceso, corroboré que efectivamente Brasil tiene 27 estados, incluyendo Brasilia, que es un Distrito Federal, y tiene 5570 municipios, los cuales están siendo referenciados como 'customer_city' en nuestro dataset. Observé que los duplicados, en este caso, están mostrando los valores correctos, con un número de 4118 municipios y 27 estados.

Conclusiones:

- Se identificaron 3345 duplicados en la columna de identificación de clientes.
- En Brasil no se les llama ciudades, sino municipios; por lo tanto, se corregió el nombre de la columna.
- Se confirmo que tenemos clientes de todos los municipios de Brasil.
- Se observo que el municipio que más clientes tiene es Sao Paulo, con un 15% del total.
- Además, Sao Paulo también es el que más clientes posee, con un 42% del total de clientes.
- Tenemos clientes en el 75% de Brasil.

FUENTE: (https://chatgpt.com/share/9822ce44-a0e4-454f-9659-64eec23b63aa)

----

### **Dataset Ordenes**

Revisamos la composición de los datos.

In [22]:
orders.head(5)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status   order_approved_at order_received_by_local_operator_date  \
0    delivered 2017-10-02 11:07:15                   2017-10-04 19:55:00   
1    delivered 2018-07-26 03:24:27                   2018-07-26 14:31:00   
2    delivered 2018-08-08 08:55:23                   2018-08-08 13:50:00   
3    delivered 2017-11-18 19:45:59                   2017-11-22 13:39:59   
4    delivered 2018-02-13 22:20:29                   2018-02-14 19:46:34   

  order_delivered_to_customer_date order_delivery_promise_date  
0              2017-10-10 21:25:13                  2017-10-18  
1              2018-08-07 15:27:45                  2018-08-13  
2              2018-08-17 18:06:29                  2018-09-04  
3              2017-12-02 00:28:42                  2017-12-15  
4              2018-02-16 18:17:02                  2018-02-26

### ``order_id``

Observar la ausencia de duplicados en nuestros registros de pedidos resalta una generación sólida de pedidos únicos, lo que confirma la integridad de nuestro sistema y la eficacia de nuestros procesos de registro.

In [23]:
orders['order_id'].duplicated().value_counts()

order_id
False    99281
Name: count, dtype: int64

### ``customer_id``

Hemos observado que no existen duplicados en los registros de identificación de clientes en relación con los pedidos.

In [24]:
orders['customer_id'].duplicated().value_counts()

customer_id
False    99281
Name: count, dtype: int64

Para determinar qué cliente realizó más pedidos, decidimos realizar un conteo y verificar el ID correspondiente. Sin embargo, observamos que no hay más de un pedido por cliente, lo que nos parece inconsistente. Nuestro dataset previo de clientes, después de eliminar duplicados, contiene un total de 96096 registros, lo que indica que tenemos esa cantidad de clientes. 

Al analizar el dataset de órdenes y realizar el proceso de contar valores para verificar si algún cliente realizó más de un pedido, notamos que ninguno se repite. Esto nos lleva a la conclusión de que los valores no están correctos.

En resumen, nuestro dataset de clientes tiene un total de 96,096 clientes, y nuestro dataset de órdenes, que no contiene más de una orden por cliente, nos muestra un total de 99281 pedidos, lo cual no concuerda, ya que ningún cliente ha realizado más de un pedido.

In [25]:
orders['customer_id'].value_counts()

customer_id
9ef432eb6251297304e76186b10a928d    1
82eb0f700864871ddc51572601014ba4    1
b794220321d905dbfd3740e2292244bd    1
12ca54ef905df73b98dcf7ccb0b92d92    1
b67d9091d904c9237c3f00382f721c5b    1
                                   ..
16be3ab0d37467d1f96d6fc30613ee57    1
6260e99309fefdfaef12d0ef8fe82e77    1
2ab60fd96df98dda774845cbc22d5468    1
becec6158cf50315fdde081c3264fdb9    1
edb027a75a1449115f6b43211ae02a24    1
Name: count, Length: 99281, dtype: int64

### ``order_status``

Observamos que tenemos 7 posibles estados para nuestras órdenes, que son:

- Entregado 
- Enviado 
- No disponible 
- Cancelado 
- Facturado 
- En proceso 
- Aprobado

In [26]:
F.cantidad_porcentaje(orders,'order_status')

Los valores de order_status:
delivered      96464
shipped         1107
unavailable      609
canceled         484
invoiced         314
processing       301
approved           2

El porcentaje que representa cada valor:
delivered      97.16
shipped         1.12
unavailable     0.61
canceled        0.49
invoiced        0.32
processing      0.30
approved        0.00


### ``order_approved_at``

Se realizó una descripción de los valores que presenta la columna 'order_approved_at', y lo que nos interesa son los valores mínimos y máximos para poder determinar el rango de fechas en el cual se recibieron pedidos. Según el dataset, nuestro primer pedido fue el 15/09/2016 y el último fue el 03/09/2018.

In [27]:
orders['order_approved_at'].describe()

count                            99281
mean     2017-12-31 18:35:24.098800384
min                2016-09-15 12:16:38
25%                2017-09-12 23:24:16
50%                2018-01-19 11:36:13
75%                2018-05-04 20:35:10
max                2018-09-03 17:40:06
Name: order_approved_at, dtype: object

### ``order_received_by_local_operator_date``

En el inicio de este proceso ETL, se corroboró que esta columna tiene valores nulos (1637). Ahora procederemos a ubicar esos registros.

In [28]:
ordenes_nulas_recibidas = orders.loc[orders['order_received_by_local_operator_date'].isnull()]
ordenes_nulas_recibidas['order_status'].value_counts()

order_status
unavailable    609
canceled       409
invoiced       314
processing     301
approved         2
delivered        2
Name: count, dtype: int64

Observamos una gran similitud en los valores encontrados en el conteo de 'order_status', siendo el único donde se presenta diferencia el de los cancelados.

Esto puede deberse a que inicialmente se recibió la orden por parte del operador local y luego el cliente la canceló. Aunque se generó la orden, nunca se entregará debido a la cancelación.

En resumen, al realizar el conteo, la columna 'order_status' indica que tenemos 484 cancelaciones, mientras que al hacerlo en la columna 'order_received_by_local_operator_date', nos muestra 409 cancelaciones. Es decir, hubo 75 pedidos cancelados después de que se generara la orden para el operador local.

In [29]:
orders['order_received_by_local_operator_date'].describe()

count                            97644
mean     2018-01-04 22:55:51.933677056
min                2016-10-08 10:34:01
25%      2017-09-15 23:45:44.750000128
50%                2018-01-24 16:38:54
75%                2018-05-08 13:40:00
max                2018-09-11 19:48:28
Name: order_received_by_local_operator_date, dtype: object

Como se hizo anteriormente, verificamos que la fecha mínima no sea anterior a la mostrada en la aprobación de la primera orden.

Efectivamente, encontramos una diferencia de aproximadamente 25 días con respecto a la primera orden aprobada. Luego, la fecha máxima debería lógicamente ser posterior a la que se muestra en la última fecha de aprobación de orden, y efectivamente es 8 días después de la última orden.

Se tomó la decisión de no eliminar estos valores que están en diferentes estados al de aprobado, dado que no nos sería de confianza tener datos en los que no tenemos registros de cancelación, no disponible y otros estados para poder analizarlos.

### ``order_delivered_to_customer_date``

En esta columna encontramos más valores nulos que en la anterior, con un total de 2819 registros. Extraeremos estos registros para revisar sus estados y sacar conclusiones.

In [30]:
ordenes_nulas_entregadas = orders.loc[orders['order_delivered_to_customer_date'].isnull()]
ordenes_nulas_entregadas['order_status'].value_counts()

order_status
shipped        1107
unavailable     609
canceled        478
invoiced        314
processing      301
delivered         8
approved          2
Name: count, dtype: int64

En este conteo observamos algunas diferencias; Una gran cantidad de pedidos están marcados como enviados (1107), pero aún no han sido entregados, por lo que figuran como NaT (Not a Time) y se interpretan como valores nulos. 

Además, nuevamente encontramos discrepancias con respecto a los pedidos cancelados. En el análisis inicial de la columna 'order_status', teníamos 484 órdenes canceladas, pero en este análisis de la columna 'order_delivered_to_customer_date', encontramos 478 canceladas. Notamos que ninguna de las dos columnas coincide con el valor del análisis de la columna 'order_status', es decir, ninguna representa en su totalidad las cancelaciones, pero sí los otros estados. Esto se debe a que tenemos 6 órdenes que fueron canceladas después de ser entregadas. Ahora buscaremos esos registros.

In [31]:
filtered_orders = orders.loc[(orders['order_status'] == 'canceled') & (~orders['order_received_by_local_operator_date'].isnull()) &  (~orders['order_delivered_to_customer_date'].isnull())]

filtered_orders

order_id                       customer_id  \
2916   1950d777989f6a877539f53795b4c3c3  1bccb206de9f0f25adc6871a1bcf77b2   
8772   dabf2b0e35b423f94618bf965fcb7514  5cdec0bb8cbdf53ffc8fdc212cd247c6   
58166  770d331c84e5b214bd9dc70a10b829d0  6c57e6119369185e575b36712766b0ef   
59231  8beb59392e21af5eb9547ae1a9938d06  bf609b5741f71697f65ce3852c5d2623   
92489  65d1e226dfaeb8cdc42f665422522d14  70fc57eeae292675927697fe03ad3ff5   
94247  2c45c33d2f9cb8ff8b1c86cc28c11c30  de4caa97afa80c8eeac2ff4c8da5b72e   

      order_status   order_approved_at order_received_by_local_operator_date  \
2916      canceled 2018-02-19 20:56:05                   2018-02-20 19:57:13   
8772      canceled 2016-10-09 13:36:58                   2016-10-13 13:36:59   
58166     canceled 2016-10-07 15:07:10                   2016-10-11 15:07:11   
59231     canceled 2016-10-09 14:34:30                   2016-10-14 22:45:26   
92489     canceled 2016-10-04 10:18:57                   2016-10-25 12:14:28   
94247     canceled 2016-10-10 10:40:49                   2016-10-14 10:40:50   

      order_delivered_to_customer_date order_delivery_promise_date  
2916               2018-03-21 22:03:51                  2018-03-09  
8772               2016-10-16 14:36:59                  2016-11-30  
58166              2016-10-14 15:07:11                  2016-11-29  
59231              2016-10-19 18:47:43                  2016-11-30  
92489              2016-11-08 10:58:34                  2016-11-25  
94247              2016-11-09 14:53:50                  2016-12-08

Como originalmente se presumía, tenemos 6 registros con cancelaciones de diferencia, los cuales ya habían sido entregados.

In [32]:
orders['order_delivered_to_customer_date'].describe()

count                            96462
mean     2018-01-14 13:16:24.206247168
min                2016-10-11 13:46:32
25%         2017-09-25 22:32:42.500000
50%                2018-02-02 19:49:47
75%         2018-05-15 23:07:37.500000
max                2018-10-17 13:22:46
Name: order_delivered_to_customer_date, dtype: object

Como se hizo anteriormente, verificamos que la fecha mínima no sea anterior a la que se muestra en la aprobación de la primera orden. Efectivamente, encontramos una diferencia de 3 días con respecto a la primera orden enviada al operador local.

Luego, la fecha máxima debería lógicamente ser posterior a la que se muestra en la última fecha de recepción de orden del operador local, y efectivamente es 36 días después de la última orden recibida por el operador local.

### ``order_delivery_promise_date``

Se procedió a crear un nuevo DataFrame en el cual se realizó la resta de la columna 'order_delivery_promise_date' (fecha prometida) y 'order_delivered_to_customer_date' (fecha de entrega) para verificar si hay retrasos, y se confirmó que sí los hay.  

Se encontraron un total de 7827 demoras con respecto a la fecha prometida.

In [33]:
nuevo_df = pd.DataFrame({ 'fecha_entrega': orders['order_delivered_to_customer_date'],
'fecha_prometida': orders['order_delivery_promise_date'] })

nuevo_df['duracion'] = nuevo_df['fecha_prometida'] - nuevo_df['fecha_entrega']
nuevo_df['duracion'] = nuevo_df['duracion'].dt.days
nuevo_df = nuevo_df.dropna(subset=['duracion']) #Se quitan los valores negativos dado que no se le puede cambiar el tipo de dato a la columna si posee valores negativos.

nuevo_df['duracion'] = nuevo_df['duracion'].astype(int)
nuevo_df['duracion'].value_counts()

entregas_tempranas = nuevo_df[nuevo_df['duracion'] > 0]['duracion'].count()
demoras = nuevo_df[nuevo_df['duracion'] < 0]['duracion'].count()

print(f"Entregas tempranas: {entregas_tempranas}")
print(f"Demoras: {demoras}")


Entregas tempranas: 87173
Demoras: 7827


In [34]:
nuevo_df['duracion'].describe()

count    96462.000000
mean        10.875391
std         10.183586
min       -189.000000
25%          6.000000
50%         11.000000
75%         16.000000
max        146.000000
Name: duracion, dtype: float64

Revisamos una descripción de la duración de los envíos, habiendo eliminado previamente los valores nulos, que suman un total de 2819 registros. Esto se considera que no modifica el valor promedio real de entregas, ya que al no tener una fecha de entrega, estos registros no aportaron datos; no se les puede restar nada a la fecha prometida de entrega.

Las conclusiones son:
- El total de registros es de 99281.
- Tenemos 96462 registros completos, lo que representa una gran cantidad con respecto al total.
- El promedio de demora en los envíos, incluyendo los outliers (valores atípicos) que sesgan un poco el valor real de las demoras, se sitúa entre 10 y 11 días.
- Se observan valores grandes de demora, como 189 días, y valores de entrega sin coherencia, como 146 días antes de lo prometido.
- Se detectan incongruencias en cuanto a la cantidad de pedidos y la cantidad de clientes.
- Tenemos un 97% de órdenes entregadas.
- La coherencia de las columnas de fechas es buena, pero se observan valores anormales.


## Exportamos archivos 🌐

Se realizaron modificaciones en el DataFrame de Clientes, por lo que será exportado nuevamente para asegurar consistencia y calidad en la parte analítica. Además, se exportará el de órdenes para pasarlo a formato CSV, al igual que el de clientes.

In [36]:
customers.to_excel('../Datasets/bc_customers_dataset_modif.xlsx',index=False)
orders.to_excel('../Datasets/bc_orders_dataset_modif.xlsx',index=False)